File: study_upload_file/keykat/210724_programmers_67257/programmers_67257.py
from itertools import permutations

# + - *, 100+200-300*400
# 1. 맨 뒤의 걸로 스플릿 --> 100+200-300, 400
# 2. 그 다음 걸로 스플릿 --> 100+200, 300(-), 400(*)
# 3. 연산 후 return --> 300, 300, 400
# 4. 재귀 --> 300-300, 400 --> 0, 400 --> 0
def calc(exp, expression):

    if exp == 1:
        return str(eval(expression))

    # +-*이면 -> * -> - -> +
    expression = expression.split(exp[-1])
    res = ''

    # 100+200-300, 400
    for e in expression:
        if '+' not in e and '-' not in e and '*' not in e:
            res = res + e + exp[-1]
        else:
            res = res + str(calc(exp[0:-1], e)) + exp[-1]
    res = res[0:-1]

    return eval(res)


def solution(expression):
    answer = []

    exp = ['+','-','*']
    exp = list(map(''.join, permutations(exp, 3)))

    for e in exp:
        answer.append(abs(calc(e, expression)))

    #print(answer)
    return max(answer)

File: study_upload_file/keykat/211214_programmers_67257/programmers_67257.kt
class Solution67257 {

    fun calcurator(c: String, a: Long, b: Long): Long {
        return if (c == "+") {
            a + b
        } else if (c == "-") {
            a - b
        } else a * b
    }
    
    fun calc(a: String, str: MutableList<String>): MutableList<String> {
        var list = mutableListOf<String>()
        var idx = 0

        while (idx <= str.size - 1) {
            if (str[idx] == a) {
                var num = list.removeAt(list.size - 1)
                list.add(calcurator(a, num.toLong(), str[idx + 1].toLong()).toString())
                idx += 2
            } else {
                list.add(str[idx])
                idx +=  1
            }
        }

        return list
    }

    fun solution(expression: String): Long {
        var list = mutableListOf<String>()
        var answer: Long = 0
        var curr = ""
        expression.forEach { 
            if (it == '+' || it == '-' || it == '*') {
                list.add(curr)
                list.add(it.toString())
                curr = ""
            } else {
                curr += it
            }
        }
        list.add(curr)

        answer = Math.max(Math.abs(calc("*", calc("+", calc("-", list))).first().toLong()), answer)
        answer = Math.max(Math.abs(calc("*", calc("-", calc("+", list))).first().toLong()), answer)
        answer = Math.max(Math.abs(calc("+", calc("*", calc("-", list))).first().toLong()), answer)
        answer = Math.max(Math.abs(calc("+", calc("-", calc("*", list))).first().toLong()), answer)
        answer = Math.max(Math.abs(calc("-", calc("*", calc("+", list))).first().toLong()), answer)
        answer = Math.max(Math.abs(calc("-", calc("+", calc("*", list))).first().toLong()), answer)
        
        return answer
    }
}